crashtracker/src/collector/crash_handler.rs

```diff
@@ -2,35 +2,227 @@
+        |f| {
+            OpenOptions::new()
+                .append(true)
+                .create(true)
+                .open(f)
+                .with_context(|| format!("Failed to open or create file: {f}"))
+        },
+    )?;
+    Ok(file.into_raw_fd())
+}
+
+/// Non-blocking child reaper
+/// * If the child process has exited, return true
+/// * If the child process cannot be found, return false
+/// * If the child is still alive, or some other error occurs, return an error Either way, after
+///   this returns, you probably don't have to do anything else.
+fn reap_child_non_blocking(pid: Pid, timeout_ms: u32) -> anyhow::Result<bool> {
+    let timeout = Duration::from_millis(timeout_ms as u64);
+    let start_time = Instant::now();
+
+    loop {
+        match waitpid(pid, Some(WaitPidFlag::WNOHANG)) {
+            Ok(WaitStatus::StillAlive) => {
+                if Instant::now().duration_since(start_time) > timeout {
+                    return Err(anyhow::anyhow!("Timeout waiting for child process to exit"));
+                }
+            }
+            Ok(_status) => {
+                return Ok(true);
+            }
+            Err(nix::Error::ECHILD) => {
+                // Non-availability of the specified process is weird, since we should have
+                // exclusive access to reaping its exit, but at the very least means there is
+                // nothing further for us to do.
+                return Ok(true);
+            }
+            _ => {
+                return Err(anyhow::anyhow!("Error waiting for child process to exit"));
+            }
+        }
+    }
+}
+
+/// Wrapper around the child process that will run the crash receiver
+fn run_receiver_child(uds_parent: RawFd, uds_child: RawFd, stderr: RawFd, stdout: RawFd) -> ! {
+    // File descriptor management
+    unsafe {
+        dup2(uds_child, 0);
+        dup2(stdout, 1);
+        dup2(stderr, 2);
+    }
+
+    // Close unused file descriptors
+    let _ = close(uds_parent);
+    let _ = close(uds_child);
+    let _ = close(stderr);
+    let _ = close(stdout);
+
+    // We've already prepared the arguments and environment variable
+    // If we've reached this point, it means we've prepared the arguments and environment variables
+    // ahead of time.  Extract them now.  This was prepared in advance in order to avoid heap
+    // allocations in the signal handler.
+    let receiver_args = RECEIVER_ARGS.load(SeqCst);
+    let (binary_path, args_ptrs, env_vars_ptrs) = unsafe {
+        let receiver_args = receiver_args.as_ref().expect("No receiver arguments");
+        (
+            &receiver_args.binary_path,
+            &receiver_args.args_ptrs,
+            &receiver_args.env_vars_ptrs,
+        )
+    };
+
+    // Print the first string in the environment variables to stderr
+    // This is a way to check that the environment variables are being passed correctly
+    println!("Crashtracker receiver environment variable: {:?}", unsafe {
```

Comment:
Where will this message be printed? Isn't there a risk of leaking some information?

Reply:
Sorry, this is wrong--debug statement.  Removed.

Reply:
This is an abomination; stricken